File: internal/service/data_svc/mall_weather_capacity_plan_test.go
package data_svc

import (
	"context"
	"errors"
	"math"
	"reflect"
	"testing"
	"time"
)

func TestBuildMallWeatherCapacityPlanEstimatesProviderAndStoragePressure(t *testing.T) {
	t.Parallel()

	plan, err := BuildMallWeatherCapacityPlan(MallWeatherCapacityPlanInput{
		MallCount: 5, ProviderQPS: 2.5, HourlySteps: 360, DailySteps: 15, LifeIndexDays: 15, AlertsPerMall: 3,
	})
	if err != nil {
		t.Fatalf("BuildMallWeatherCapacityPlan() error=%v", err)
	}
	if plan.WeatherV26ProviderRequestsPerDay != 840 || plan.LifeIndexV3ProviderRequestsPerDay != 0 ||
		plan.ProviderRequests != 840 || plan.ProviderDrainSeconds != 336 ||
		plan.MinimumQPSForOneHourDrain != float64(840)/3600 {
		t.Fatalf("provider plan=%+v", plan)
	}
	if plan.TotalDatabaseRows != 2570 || plan.TotalDatabaseBatches != 35 || plan.FeishuBatchRows != defaultMallWeatherFeishuBatchRows ||
		plan.TotalFeishuBatches != 16 {
		t.Fatalf("plan totals=%+v", plan)
	}
	wantDatasets := []MallWeatherCapacityDatasetPlan{
		{Kind: "realtime", Rows: 5, DatabaseBatches: 5, FeishuBatches: 1},
		{Kind: "minutely", Rows: 600, DatabaseBatches: 5, FeishuBatches: 3},
		{Kind: "hourly", Rows: 1800, DatabaseBatches: 10, FeishuBatches: 9},
		{Kind: "daily", Rows: 75, DatabaseBatches: 5, FeishuBatches: 1},
		{Kind: "alerts", Rows: 15, DatabaseBatches: 5, FeishuBatches: 1},
		{Kind: "life_indices", Rows: 75, DatabaseBatches: 5, FeishuBatches: 1},
	}
	if !reflect.DeepEqual(plan.Datasets, wantDatasets) {
		t.Fatalf("datasets=%+v want %+v", plan.Datasets, wantDatasets)
	}
}

func TestBuildMallWeatherCapacityPlanUsesConfiguredFeishuBatchRows(t *testing.T) {
	t.Parallel()

	plan, err := BuildMallWeatherCapacityPlan(MallWeatherCapacityPlanInput{
		MallCount: 2, ProviderQPS: 1, HourlySteps: 24, DailySteps: 1, LifeIndexDays: 1, AlertsPerMall: 0, FeishuBatchRows: 50,
	})
	if err != nil {
		t.Fatalf("BuildMallWeatherCapacityPlan() error=%v", err)
	}
	if plan.FeishuBatchRows != 50 || plan.TotalFeishuBatches != 9 {
		t.Fatalf("plan=%+v", plan)
	}
}

func TestBuildMallWeatherCapacityPlanRejectsInvalidInputs(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name   string
		mutate func(*MallWeatherCapacityPlanInput)
	}{
		{name: "missing mall count", mutate: func(input *MallWeatherCapacityPlanInput) { input.MallCount = 0 }},
		{name: "missing qps", mutate: func(input *MallWeatherCapacityPlanInput) { input.ProviderQPS = 0 }},
		{name: "nan qps", mutate: func(input *MallWeatherCapacityPlanInput) { input.ProviderQPS = math.NaN() }},
		{name: "infinite qps", mutate: func(input *MallWeatherCapacityPlanInput) { input.ProviderQPS = math.Inf(1) }},
		{name: "too many hourly steps", mutate: func(input *MallWeatherCapacityPlanInput) { input.HourlySteps = 361 }},
		{name: "too many daily steps", mutate: func(input *MallWeatherCapacityPlanInput) { input.DailySteps = 16 }},
		{name: "too many life days", mutate: func(input *MallWeatherCapacityPlanInput) { input.LifeIndexDays = 16 }},
		{name: "too many alerts", mutate: func(input *MallWeatherCapacityPlanInput) { input.AlertsPerMall = 257 }},
		{name: "too many feishu rows", mutate: func(input *MallWeatherCapacityPlanInput) { input.FeishuBatchRows = 501 }},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			input := MallWeatherCapacityPlanInput{
				MallCount: 1, ProviderQPS: 1, HourlySteps: 1, DailySteps: 1, LifeIndexDays: 1,
			}
			test.mutate(&input)
			if _, err := BuildMallWeatherCapacityPlan(input); !errors.Is(err, ErrMallWeatherInvalidCapacityPlan) {
				t.Fatalf("BuildMallWeatherCapacityPlan() error=%v want %v", err, ErrMallWeatherInvalidCapacityPlan)
			}
		})
	}
}

func TestMallWeatherCapacityPlanServiceRequiresConfigPermission(t *testing.T) {
	t.Parallel()

	now := time.Date(2026, 7, 23, 9, 0, 0, 0, time.UTC)
	permissions := &recordingMallPermissionChecker{allowed: map[string]bool{
		PermissionWeatherConfigManage: true,
	}}
	service, err := newMallWeatherCapacityPlanService(permissions, func() time.Time { return now })
	if err != nil {
		t.Fatalf("newMallWeatherCapacityPlanService() error=%v", err)
	}
	plan, err := service.Plan(context.Background(), 17, MallWeatherCapacityPlanInput{
		MallCount: 1, ProviderQPS: 1, HourlySteps: 24, DailySteps: 1, LifeIndexDays: 1,
	})
	if err != nil {
		t.Fatalf("Plan() error=%v", err)
	}
	if plan.ProviderRequests == 0 || !reflect.DeepEqual(permissions.requested, []string{PermissionWeatherConfigManage}) {
		t.Fatalf("plan=%+v permissions=%v", plan, permissions.requested)
	}
}

func TestMallWeatherCapacityPlanServiceRejectsUnauthorizedActor(t *testing.T) {
	t.Parallel()

	service, err := newMallWeatherCapacityPlanService(
		&recordingMallPermissionChecker{allowed: map[string]bool{}},
		time.Now,
	)
	if err != nil {
		t.Fatalf("newMallWeatherCapacityPlanService() error=%v", err)
	}
	_, err = service.Plan(context.Background(), 17, MallWeatherCapacityPlanInput{
		MallCount: 1, ProviderQPS: 1, HourlySteps: 24, DailySteps: 1, LifeIndexDays: 1,
	})
	if !errors.Is(err, ErrMallForbidden) {
		t.Fatalf("Plan() error=%v, want ErrMallForbidden", err)
	}
	if _, err = service.Plan(context.Background(), 0, MallWeatherCapacityPlanInput{}); !errors.Is(err, ErrMallForbidden) {
		t.Fatalf("Plan() anonymous error=%v, want ErrMallForbidden", err)
	}
}

File: web/tests/mallWeatherCapacityPlan.test.mjs
import assert from 'node:assert/strict'
import test from 'node:test'

import { mallWeatherCapacityPlanPath, parseMallWeatherCapacityPlan } from '../.test-dist/mallWeatherCapacityPlan.js'

const input = { mallCount: '5', providerQps: '2.5', hourlySteps: '360', dailySteps: '15', lifeIndexDays: '15', alertsPerMall: '3', feishuBatchRows: '200' }

test('builds the exact bounded capacity-plan query', () => {
  assert.equal(mallWeatherCapacityPlanPath(input), '/v1/mall-weather/capacity-plan?mallCount=5&providerQps=2.5&hourlySteps=360&dailySteps=15&lifeIndexDays=15&alertsPerMall=3&feishuBatchRows=200')
  assert.equal(mallWeatherCapacityPlanPath({ ...input, mallCount: '100000', providerQps: '0.0001', hourlySteps: '1', dailySteps: '1', lifeIndexDays: '1', alertsPerMall: '0', feishuBatchRows: '1' }), '/v1/mall-weather/capacity-plan?mallCount=100000&providerQps=0.0001&hourlySteps=1&dailySteps=1&lifeIndexDays=1&alertsPerMall=0&feishuBatchRows=1')
  for (const invalid of [
    { mallCount: '0' }, { mallCount: '100001' }, { providerQps: '0' }, { providerQps: '10001' },
    { hourlySteps: '0' }, { hourlySteps: '361' }, { dailySteps: '0' }, { dailySteps: '16' },
    { lifeIndexDays: '0' }, { lifeIndexDays: '16' }, { alertsPerMall: '-1' }, { alertsPerMall: '257' },
    { feishuBatchRows: '0' }, { feishuBatchRows: '501' },
  ]) assert.throws(() => mallWeatherCapacityPlanPath({ ...input, ...invalid }), /invalid capacity plan input/)
})

test('parses only complete safe capacity plan responses', () => {
  const payload = { code: 0, data: {
    mallCount: 5, providerQps: 2.5, weatherV26ProviderRequestsPerDay: 840, lifeIndexV3ProviderRequestsPerDay: 0,
    providerRequests: 840, providerDrainSeconds: 336, minimumQpsForOneHourDrain: 840 / 3600,
    totalDatabaseRows: 2570, totalDatabaseBatches: 35, feishuBatchRows: 200, totalFeishuBatches: 16,
    datasets: [{ kind: 'hourly', rows: 1800, databaseBatches: 10, feishuBatches: 9 }],
  } }
  assert.deepEqual(parseMallWeatherCapacityPlan(payload), payload.data)
  assert.equal(parseMallWeatherCapacityPlan({ ...payload, data: { ...payload.data, datasets: [{ ...payload.data.datasets[0], rows: -1 }] } }), null)
  assert.equal(parseMallWeatherCapacityPlan({ ...payload, data: { ...payload.data, providerQps: 0 } }), null)
  assert.equal(parseMallWeatherCapacityPlan({ ...payload, code: 200 }), null)
  assert.equal(parseMallWeatherCapacityPlan({ ...payload, data: { ...payload.data, totalFeishuBatches: undefined } }), null)
  assert.equal(parseMallWeatherCapacityPlan({ ...payload, data: { ...payload.data, providerDrainSeconds: Number.POSITIVE_INFINITY } }), null)
  assert.equal(parseMallWeatherCapacityPlan({ ...payload, data: { ...payload.data, datasets: [{ kind: 'hourly', rows: 1, databaseBatches: 1 }] } }), null)
})
